# Додаткове завдання з pandas
Вітаю! В цьому завданні будемо працювати з даними з опитування щодо зарплат ІТ спеціалістів на рексурсі DOU від грудня 2022 року. Впевнена, це буде цікаве дослідження, а заразом і практика pandas :)
Цей аналіз може бути корисним людині, яка обирає напрям роботи в ІТ і хоче зрозуміти тенденції. За якісного виконання цього аналізу, цей проєкт можна сміливо додавати у Ваше портфоліо!

1. Зчитайте дані з файлу `2022_dec_raw.csv`. Увага! Тут необхідно використати аргумент `pd.read_csv` `sep=';'`, бо розділювач в даних - крапка з комою.

In [2]:
import pandas as pd
df = pd.read_csv('/Users/alinagajdun/Documents/python for DA/data/2022_dec_raw.csv', sep=';')

2. Які колонки мають дані?

In [4]:
df.columns

Index(['Timestamp', 'Ви працюєте зараз в ІТ?', 'Де ви зараз живете? ',
       'В якій області ви зараз живете?',
       'Зарплата у $$$ за місяць, лише ставка після сплати податків',
       'Ваша стать', 'Ваш вік', 'Яка у вас освіта?',
       'Чи є у вас науковий ступінь? ', 'Знання англійської мови',
       'Оберіть вашу посаду', 'Ваш тайтл', 'Ваша спеціалізація',
       'Вкажіть вашу спеціалізацію', 'Оберіть вашу спеціалізацію',
       'Ваша посада', 'Оберіть вашу посаду2', 'Ваша посада3', 'Ваша посада4',
       'Ваша посада5', 'Ваша посада6', 'Ваша посада7',
       'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
       'Основна мова програмування', 'Вкажіть вашу основну спеціалізацію',
       'В якій сфері працюєте?', 'Тип компанії',
       'Загальний стаж роботи за спеціальністю'],
      dtype='object')

3. Скільки рядків та колонок мають дані? Скільки пропущених значень має кожна з колонок?

In [687]:
print(f'Дані мають розмір {df.shape}')
print(f'Пропущені значення в даних: \n{df.isnull().sum()}')

Дані мають розмір (13276, 29)
Пропущені значення в даних: 
Timestamp                                                                       506
Ви працюєте зараз в ІТ?                                                           0
Де ви зараз живете?                                                               0
В якій області ви зараз живете?                                                1748
Зарплата у $$$ за місяць, лише ставка після сплати податків                       0
Ваша стать                                                                        0
Ваш вік                                                                           0
Яка у вас освіта?                                                                 0
Чи є у вас науковий ступінь?                                                    111
Знання англійської мови                                                           0
Оберіть вашу посаду                                                               0
Ваш тайтл        

4. Які колонки мають всі значення заповненими (не мають дожного None значення)? Виведіть програмно список цих колонок.

In [8]:
columns_without_null = df.columns[df.notnull().all()]
print(columns_without_null)

Index(['Timestamp', 'Ви працюєте зараз в ІТ?', 'Де ви зараз живете? ',
       'Ваша стать', 'Ваш вік', 'Яка у вас освіта?', 'Знання англійської мови',
       'Оберіть вашу посаду', 'Ваш тайтл', 'В якій сфері працюєте?',
       'Тип компанії', 'Загальний стаж роботи за спеціальністю'],
      dtype='object')


5. Який тип мають колонки в наборі даних? Які з колонок типу object логічно на ваш погляд перетворити на числові колонки?

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13315 entries, 0 to 13314
Data columns (total 28 columns):
 #   Column                                                                      Non-Null Count  Dtype 
---  ------                                                                      --------------  ----- 
 0   Timestamp                                                                   13315 non-null  object
 1   Ви працюєте зараз в ІТ?                                                     13315 non-null  object
 2   Де ви зараз живете?                                                         13315 non-null  object
 3   В якій області ви зараз живете?                                             11538 non-null  object
 4   Зарплата у $$$ за місяць, лише ставка після сплати податків                 13288 non-null  object
 5   Ваша стать                                                                  13315 non-null  object
 6   Ваш вік                                               

Я би перетворила тип даних наступних колонок:
- Зарплата у $$$ за місяць;
- Ваш вік;
- Загальний стаж роботи за спеціальністю;


6. Трансформуйте колонки 'Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків' у тип даних float. УВАГА! Перед трансформацією подивіться, які дані містять ці колонки. Аби успішно трансформувати тип даних, треба спочатку привести рядки до того виду, який буде підходити, аби потім рядок зробити типом float.

In [12]:
# перевіримо які дані містить наш датасет
print(f'Колонка "Ваш вік містить значення": {df["Ваш вік"].unique()}'),
print(f'Колонка "Зарплата у $$$ за місяць" містить значення: {df["Зарплата у $$$ за місяць, лише ставка після сплати податків"].unique()}')

Колонка "Ваш вік містить значення": ['29' '36' '34' '23' '33' '31' '30' '24' '27' '17' '25' '20' '26' '22'
 '19' '21' '18' '32' '35' '28' '37' '43' '44' '39' '50' '40' '38' '41'
 '45' '42' '49' '46' '47' '48' '55' '63' '58' '53' '54' '51' '56' '60'
 '16' '57' '29,5' '52' '14']
Колонка "Зарплата у $$$ за місяць" містить значення: ['4000' nan '2500' '500' '2800' '5500' '1400' '1200' '650' '6000' '6600'
 '850' '1000' '1500' '3000' '800' '700' '1100' '7500' '10600' '3300'
 '5000' '2400' '1800' '2100' '4500' '2600' '3600' '3525' '1600' '3200'
 '570' '6500' '4200' '1700' '5320' '5300' '1490' '1250' '2700' '4700'
 '1300' '400' '2000' '1577' '10000' '550' '5600' '950' '3550' '4750'
 '2200' '1850' '5200' '3040' '5100' '2250' '6450' '3500' '1900' '4400'
 '2865' '5350' '3700' '2850' '6150' '680' '2169' '1350' '2222' '2185'
 '2750' '3760' '2300' '200' '4275' '7700' '978' '1710' '1170' '3100'
 '5700' '1863' '2900' '3135' '3400' '4300' '730' '8500' '2240' '7000'
 '3650' '5400' '4800' '3800' '6700' '

In [22]:
# очистимо дані від пропущених значень та зайвих символів

def clean_data(df, args):
    for col in args:
        df[col] = df[col].replace(',', '.')
        df = df.dropna(subset= [col])
    return df

# перетворення даних до типу float

def transform_to_float(df, args):
    for col in args:
        df[col] = pd.to_numeric(df[col], errors = 'coerce')
    return df  

In [24]:
# застосуємо зміни

df = clean_data(df, ['Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків'])
df = transform_to_float(df, ['Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків'])

# переглянемо нові дані

print(f'Колонка "Ваш вік містить значення": {df["Ваш вік"].unique()}'),
print(f'Колонка "Зарплата у $$$ за місяць" містить значення: {df["Зарплата у $$$ за місяць, лише ставка після сплати податків"].unique()}') 

Колонка "Ваш вік містить значення": [29. 34. 23. 33. 31. 30. 24. 27. 25. 20. 26. 19. 21. 22. 36. 18. 32. 35.
 28. 37. 43. 44. 39. 50. 40. 38. 41. 45. 42. 49. 46. 47. 48. 55. 63. 58.
 53. 17. 54. 51. 56. 60. 16. 57. 52. 14.]
Колонка "Зарплата у $$$ за місяць" містить значення: [ 4000.  2500.   500.  2800.  5500.  1400.  1200.   650.  6000.  6600.
   850.  1000.  1500.  3000.   800.   700.  1100.  7500. 10600.  3300.
  5000.  2400.  1800.  2100.  4500.  2600.  3600.  3525.  1600.  3200.
   570.  6500.  4200.  1700.  5320.  5300.  1490.  1250.  2700.  4700.
  1300.   400.  2000.  1577. 10000.   550.  5600.   950.  3550.  4750.
  2200.  1850.  5200.  3040.  5100.  2250.  6450.  3500.  1900.  4400.
  2865.  5350.  3700.  2850.  6150.   680.  2169.  1350.  2222.  2185.
  2750.  3760.  2300.   200.  4275.  7700.   978.  1710.  1170.  3100.
  5700.  1863.  2900.  3135.  3400.  4300.   730.  8500.  2240.  7000.
  3650.  5400.  4800.  3800.  6700.  2920.  4100.  5800.   750.  2811.
   330.   970

In [26]:
# Перевіримо чи після трансформації даних маємо пусті значення
df['Ваш вік'].isnull().sum(), df['Зарплата у $$$ за місяць, лише ставка після сплати податків'].isnull().sum() 

(0, 0)

7. Це завдання з зірочкою! Нам треба транфсформувати колонку Timestamp у тип даних `datetime64[ns]`. Це не так просто як здається на першйий погляд :) Але це реальна ситуація, яка може трапитись при зборі даних в роботі аналітиком.
Я запропоную один зі способів виконати завдання, але ви можете написати по-своєму. Головне - отримати колонку - потрібного типу.
Пропоную наступний алгоритм дій:
    1. Подивитись, які дані містяться в цій колонці. Виведіть перші і останні 5 записів. Чи однаковий формат цих даних? (спойлер: ні, бо інакше все було б просто :))
    2. Нам треба враховуючи різний формат даних перетворити рядок у колонці `Timestamp` на тип `datetime64[ns]`. Як це зробити? Для перетворення рядка на `datetime64` пропоную скористатись `datetime.datetime.strptime`.
    Для цього треба імпортувати бібліотеку `datetime`. Я пропоную написати функцію для обробки кожного окремого рядка, яка враховує довжину рядка і на основі цього застосовує різний формат даних дати.
    Як прописувати формат даних можна переглянути [тут](https://www.w3schools.com/python/python_datetime.asp).
    3. Коли функція для обробки кожного окремого рядка готова, застосуйте її для всіх даних з допомогою методу `apply`.

In [28]:
df.Timestamp.iloc[:5], df.Timestamp.iloc[-5:]

(0    11.30.2022 10:26:37
 2    11.30.2022 10:26:58
 3    11.30.2022 10:27:31
 4    11.30.2022 10:27:34
 5    11.30.2022 10:27:38
 Name: Timestamp, dtype: object,
 13310    01.09.2023 11:01
 13311    01.09.2023 11:14
 13312    01.09.2023 11:19
 13313    01.09.2023 11:28
 13314    01.09.2023 11:38
 Name: Timestamp, dtype: object)

In [30]:
# варіант перетворення через пандас

df['DateTime'] = pd.to_datetime(df['Timestamp'])
df['DateTime'].iloc[:5], df['DateTime'].iloc[-5:]

(0   2022-11-30 10:26:37
 2   2022-11-30 10:26:58
 3   2022-11-30 10:27:31
 4   2022-11-30 10:27:34
 5   2022-11-30 10:27:38
 Name: DateTime, dtype: datetime64[ns],
 13310   2023-01-09 11:01:00
 13311   2023-01-09 11:14:00
 13312   2023-01-09 11:19:00
 13313   2023-01-09 11:28:00
 13314   2023-01-09 11:38:00
 Name: DateTime, dtype: datetime64[ns])

In [32]:
# Перетворення через datetime

from datetime import datetime

def convert_to_data(timestamp):
    try:
        # Перевіряємо, чи є дата в форматі 'MM.DD.YYYY HH:MM:SS'
        if '.' in timestamp.split(' ')[0] and len(timestamp.split(' ')[0].split('.')) == 3:
            # Міняємо місцями місяць і день
            month, day, year = timestamp.split(' ')[0].split('.')
            timestamp = f"{day}.{month}.{year} {timestamp.split(' ')[1]}"  # Міняємо місцями день і місяць
        
        # Перевіряємо на формат 'DD.MM.YYYY HH:MM:SS'
        if len(timestamp.split(' ')[1]) == 8:  # Це означає, що є секунди
            return datetime.strptime(timestamp, '%d.%m.%Y %H:%M:%S')
        else:  # Це формат без секунд, 'DD.MM.YYYY HH:MM'
            return datetime.strptime(timestamp, '%d.%m.%Y %H:%M')
        
    except ValueError:
            # Якщо не вдалося перетворити, повертаємо NaT
        return pd.NaT

In [40]:
df['Timestamp'] = df['Timestamp'].apply(convert_to_data)
df['Timestamp']

0       2022-11-30 10:26:37
2       2022-11-30 10:26:58
3       2022-11-30 10:27:31
4       2022-11-30 10:27:34
5       2022-11-30 10:27:38
                ...        
13310   2023-01-09 11:01:00
13311   2023-01-09 11:14:00
13312   2023-01-09 11:19:00
13313   2023-01-09 11:28:00
13314   2023-01-09 11:38:00
Name: Timestamp, Length: 13276, dtype: datetime64[ns]

8. Для колонок типу даних object (які такими лишились після всіх наших трансформацій) виведіть кількість унікальних значень.

In [139]:
# Фільтруємо тільки колонки з типом 'object'
filtered = df.select_dtypes(include='object')

# Рахуємо унікальні значення для кожної колонки
for col in filtered.columns:
    print(f"Column: {col}")
    print(filtered[col].value_counts())
    print("\n")

Column: Ви працюєте зараз в ІТ?
Так, працюю full-time      12609
Ні, тимчасово не працюю      378
Так, працюю part-time        149
Так, я фрилансер             115
Інше                          25
Name: Ви працюєте зараз в ІТ?, dtype: int64


Column: Де ви зараз живете? 
В Україні                                                      11577
За кордоном (переїхали через війну, але планую повернутися)     1699
Name: Де ви зараз живете? , dtype: int64


Column: В якій області ви зараз живете?
Київ чи область                    4408
Львів чи область                   2351
Дніпро чи область                   751
Івано-Франківськ чи область         478
Вінниця чи область                  446
Одеса чи область                    443
Харків чи область                   355
Полтава чи область                  252
Черкаси чи область                  232
Хмельницький чи область             228
Тернопіль чи область                223
Чернівці чи область                 211
Ужгород чи Закарпатська обл

In [131]:
# Ініціалізуємо список для зберігання унікальних значень об'єктних колонок
unique_objects = []

# Перебираємо всі колонки у DataFrame
for col in df.columns:
    # Перевіряємо, чи тип даних колонки - об'єкт (строковий)
    if df[col].dtype == 'object':
        # Рахуємо унікальні значення і зберігаємо їх як словник
        value_counts = df[col].value_counts().to_dict()
        # Додаємо колонку та її унікальні значення до списку
        unique_objects.append((col, value_counts))

# Створюємо DataFrame для зручного відображення результатів
unique_objects_df = pd.DataFrame(unique_objects, columns=['Column Name', 'Unique Values'])

# Налаштування для відображення повного тексту у колонках DataFrame
pd.set_option('display.max_colwidth', None)

# Виводимо DataFrame
display(unique_objects_df)

Column Name  \
0                                                      Ви працюєте зараз в ІТ?   
1                                                         Де ви зараз живете?    
2                                              В якій області ви зараз живете?   
3                                                                   Ваша стать   
4                                                            Яка у вас освіта?   
5                                                Чи є у вас науковий ступінь?    
6                                                      Знання англійської мови   
7                                                          Оберіть вашу посаду   
8                                                                    Ваш тайтл   
9                                                           Ваша спеціалізація   
10                                                  Вкажіть вашу спеціалізацію   
11                                                  Оберіть вашу спеціалізацію   
12                                                                 Ваша посада   
13                                                        Оберіть вашу посаду2   
14                                                                Ваша посада3   
15                                                                Ваша посада4   
16                                                                Ваша посада5   
17                                                                Ваша посада6   
18                                                                Ваша посада7   
19  Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?   
20                                                  Основна мова програмування   
21                                          Вкажіть вашу основну спеціалізацію   
22                                                      В якій сфері працюєте?   
23                                                                Тип компанії   
24                                      Загальний стаж роботи за спеціальністю   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              

9. Для колонок типу даних object виведіть топ 10 найбільш частих значень з кількістю рядків, в яких це значення зустрічається. Зробіть це для кожної колонки окремо. Наприклад, в циклі.  
Що цікавого бачите? Які висновки можна зробити? Що особисто Вас вразило?

In [184]:
# Отримає топ 10 найчастіших значень для кожної колонки
n = 10
for col in filtered.columns:
    top_frequency = filtered[col].value_counts(normalize=False)[:n]  # Абсолютні числа
    top_percentage = filtered[col].value_counts(normalize=True)[:n] * 100  # Відсотки
    
    # Створюємо DataFrame
    df_top = pd.DataFrame({
        'Value': top_frequency.index,
        'Frequency': top_frequency.values,
        'Percentage': top_percentage.values
        })
    
    display(f"Column: {col}",df_top)

'Column: Ви працюєте зараз в ІТ?'

Value  Frequency  Percentage
0    Так, працюю full-time      12609   94.975896
1  Ні, тимчасово не працюю        378    2.847243
2    Так, працюю part-time        149    1.122326
3         Так, я фрилансер        115    0.866225
4                     Інше         25    0.188310

'Column: Де ви зараз живете? '

Value  Frequency  \
0                                                    В Україні      11577   
1  За кордоном (переїхали через війну, але планую повернутися)       1699   

   Percentage  
0   87.202471  
1   12.797529

'Column: В якій області ви зараз живете?'

Value  Frequency  Percentage
0              Київ чи область       4408   38.237335
1             Львів чи область       2351   20.393824
2            Дніпро чи область        751    6.514573
3  Івано-Франківськ чи область        478    4.146426
4           Вінниця чи область        446    3.868841
5             Одеса чи область        443    3.842817
6            Харків чи область        355    3.079459
7           Полтава чи область        252    2.185982
8           Черкаси чи область        232    2.012491
9      Хмельницький чи область        228    1.977793

'Column: Ваша стать'

Value  Frequency  Percentage
0  Чоловік      10190   76.755047
1    Жінка       3086   23.244953

'Column: Яка у вас освіта?'

Value  Frequency  Percentage
0  Вища (бакалавр, спеціаліст, магістр)      10616   79.963845
1                              Дві вищі        997    7.509792
2                       Ще студент вишу        807    6.078638
3                               Середня        447    3.366978
4                    Середня спеціальна        409    3.080747

'Column: Чи є у вас науковий ступінь? '

Value  Frequency  Percentage
0                Немає      12941   98.298519
1  Кандидат наук / PhD        216    1.640714
2          Доктор наук          8    0.060767

'Column: Знання англійської мови'

Value  Frequency  Percentage
0  Upper-Intermediate       5522   41.593854
1        Intermediate       4462   33.609521
2            Advanced       1614   12.157276
3    Pre-Intermediate       1399   10.537813
4          Elementary        279    2.101537

'Column: Оберіть вашу посаду'

Value  \
0           Software Engineer (Junior, Middle, Senior, Team/Tech Lead, Architect)   
1                   QA Engineer (Junior, Middle, Senior, Team/Tech Lead, Manager)   
2  Project/Product/Program Manager, Product Owner, Scrum Master, Delivery Manager   
3                                                                    HR/Recruiter   
4                    Designer / Artist (Graphic, UX/UI, Game, Artist, Sound etc.)   
5                                                Analyst (Business, Data, System)   
6                                                                     DevOps, SRE   
7                                                      Marketing, SEO, Copywriter   
8                                        Data Science, Machine Learning, Big Data   
9                                                     Game Developer / Programmer   

   Frequency  Percentage  
0       6743   50.790901  
1       2362   17.791503  
2        821    6.184092  
3        559    4.210606  
4        485    3.653209  
5        485    3.653209  
6        352    2.651401  
7        304    2.289846  
8        288    2.169328  
9        160    1.205182

'Column: Ваш тайтл'

Value  Frequency  Percentage
0          Middle       4691   35.334438
1          Senior       3393   25.557397
2          Junior       2313   17.422416
3       Team Lead        881    6.636035
4    Немає тайтлу        568    4.278397
5       Tech Lead        428    3.223863
6         Manager        373    2.809581
7            Head        250    1.883097
8  Intern/Trainee        240    1.807773
9       Architect        139    1.047002

'Column: Ваша спеціалізація'

Value  Frequency  Percentage
0                     Programmer         89   55.974843
1            Gameplay Programmer         42   26.415094
2              Engine Programmer          8    5.031447
3                .NET Programmer          6    3.773585
4                  Ul Programmer          4    2.515723
5  Rendering/Graphics Programmer          4    2.515723
6                  Al Programmer          3    1.886792
7                           Інше          3    1.886792

'Column: Вкажіть вашу спеціалізацію'

Value  Frequency  Percentage
0       Manual QA       1467   62.108383
1   Automation QA        530   22.438611
2      General QA        353   14.944962
3  Performance QA         12    0.508044

'Column: Оберіть вашу спеціалізацію'

Value  Frequency  Percentage
0            UI/UX Designer        200   41.237113
1          Product Designer         60   12.371134
2                 3D Artist         40    8.247423
3                 2D Artist         35    7.216495
4  Graphic / Video Designer         33    6.804124
5             Game Designer         20    4.123711
6                  Designer         15    3.092784
7                  Animator         14    2.886598
8                     Other         12    2.474227
9              Web Designer         11    2.268041

'Column: Ваша посада'

Value  Frequency  Percentage
0                          Other         44   14.473684
1              Marketing Manager         44   14.473684
2                 SEO Specialist         34   11.184211
3      Digital Marketing Manager         32   10.526316
4                     Copywriter         27    8.881579
5  Chief Marketing Officer (CMO)         22    7.236842
6         Employer Brand Manager         20    6.578947
7                    SMM Manager         20    6.578947
8                     PR Manager         16    5.263158
9      Product Marketing Manager         11    3.618421

'Column: Оберіть вашу посаду2'

Value  Frequency  Percentage
0                           Recruiter        201   35.957066
1                          HR Manager        131   23.434705
2          Talent Acquisition Manager         52    9.302326
3           Talent Researcher/Sourcer         51    9.123435
4                         HR Director         45    8.050089
5  People partner / Happiness manager         40    7.155635
6                 HR Business Partner         39    6.976744

'Column: Ваша посада3'

Value  Frequency  Percentage
0   Project Manager        443   53.958587
1   Product Manager        135   16.443362
2  Delivery Manager         88   10.718636
3     Product Owner         67    8.160780
4      Scrum Master         42    5.115713
5   Program Manager         29    3.532278
6  Business Manager         17    2.070646

'Column: Ваша посада4'

Value  Frequency  Percentage
0  Business Analyst        289   59.587629
1      Data Analyst        106   21.855670
2   Product Analyst         34    7.010309
3    System Analyst         32    6.597938
4        BI Analyst         24    4.948454

'Column: Ваша посада5'

Value  Frequency  Percentage
0  Data Engineer / Big Data Engineer        104   36.111111
1                     Data Scientist         93   32.291667
2          Machine Learning Engineer         46   15.972222
3           Computer Vision Engineer         21    7.291667
4                  Research Engineer         15    5.208333
5                        BI Engineer          9    3.125000

'Column: Ваша посада6'

Value  Frequency  Percentage
0            DevOps Engineer        325   92.329545
1  Site Reliability Engineer         27    7.670455

'Column: Ваша посада7'

Value  Frequency  Percentage
0                      CTO         41   35.344828
1  Director of Engineering         33   28.448276
2       CEO, (co-) founder         28   24.137931
3         Program Director         14   12.068966

'Column: Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?'

Value  Frequency  Percentage
0  Ні, не використовую       2751   56.500308
1    Так, використовую       2118   43.499692

'Column: Основна мова програмування'

Value  Frequency  Percentage
0  JavaScript       1692   18.741693
1        Java       1236   13.690740
2      Python       1186   13.136907
3   C# / .NET       1174   13.003988
4  TypeScript       1174   13.003988
5         PHP        624    6.911830
6      Kotlin        285    3.156845
7         C++        251    2.780239
8       Swift        221    2.447940
9         SQL        200    2.215330

'Column: Вкажіть вашу основну спеціалізацію'

Value  Frequency  Percentage
0             Back-end  розробка       2833   31.383627
1            Full Stack розробка       1622   17.968317
2            Front-end  розробка       1621   17.957239
3                             QA       1078   11.941952
4               Mobile  розробка        663    7.344633
5  Робота з даними, аналіз даних        558    6.181456
6                         DevOps        307    3.400908
7                           Інше        293    3.245818
8                       Embedded         48    0.531738
9                       embedded          2    0.022156

'Column: В якій сфері працюєте?'

Value  Frequency  Percentage
0                  Інша       1441   10.854173
1            E-commerce       1238    9.325098
2     Fintech / Banking        938    7.065381
3  Medtech / Healthcare        545    4.105152
4               GameDev        449    3.382043
5              Gambling        352    2.651401
6    Edtech / Education        249    1.875565
7               Telecom        232    1.747514
8                Mobile        199    1.498945
9  Adtech / Advertising        191    1.438686

'Column: Тип компанії'

Value  Frequency  Percentage
0  Аутсорсингова       5578   42.015667
1     Продуктова       4926   37.104550
2  Аутстафінгова       1761   13.264538
3        Стартап        537    4.044893
4           Інша        354    2.666466
5        Фріланс        120    0.903887

'Column: Загальний стаж роботи за спеціальністю'

Value  Frequency  Percentage
0             3 роки       1600   12.051823
1             4 роки       1444   10.876770
2             2 роки       1421   10.703525
3            5 років       1349   10.161193
4           1,5 року       1050    7.909009
5            6 років        978    7.366677
6              1 рік        922    6.944863
7            7 років        791    5.958120
8           10 років        594    4.474239
9  15 і більше років        586    4.413980

- Мене вразило що не зважаючи на війну в країні, в нас працює full-time майже 95% опитувальних, і відповідно платять податки і підтримують економіку в такий складний час, це свідчить про високий рівень адаптивності й стійкості спеціалістів IT-сфери в Україні.
- Дніпро та область входить в топ 3 де зараз проживаються опитувальні, а це небезпечний район близько до лінії зіткнення, це підкреслює готовність бізнесів і людей функціонувати навіть у складних умовах, підтримуючи стабільність у регіонах, близьких до зони бойових дій.
- В цій статистиці участь чоловіків в 3 рази більше ніж жінок, що може свідчити про гендерний дисбаланс в IT, що відкриває можливості для залучення й підтримки жінок у цій галузі.
- В нас в країні користується популярністю вища освіта, і згідно статистики майже 80% опитувальних мають вищу освіту - це може бути пов’язано з високими вимогами до кваліфікації в IT-секторі.
- 41% опитувальних мають доволі хороший рівень володіння англійської, але якщо ми розглядаємо сектор IT, то на мою думку цього мало, оскільки і документація, і переважна більшість IT компаній працюють з іноземними замовниками.І хоча рівні Upper-Intermediate і Advanced разом становить 53%, що є хорошим показником, але для конкурентності на міжнародному ринку варто сприяти покращенню рівня володіння мовою.
- 50% людей працюють в спеціальності Software Engineer, що показує фокус ринку на технічних спеціальностях, але також вказує на те, що інші спеціалізації можуть бути менш заповненими, як-от DevOps або Data Science.
- В нашій статистиці найчастіший тайтл це мідл -це свідчить про зрілий рівень ринку праці, де велика частина спеціалістів уже має достатній досвід для виконання ключових завдань.
- Дуже мало серед опитувальних Al Programmer - менше 2%, а цей напрям популярний, отже тим хто хоче і буде входити в IT можна займати цю нішу і стати компетентним і дуже затребуваним спеціалістом, враховуючи зростання популярності AI у світі.
- Більше половини Manual QA, а це означає що ринок тестувальників може бути трохи перенасичений, і я би не радила починати свій шлях з вибору ції спеціалізації. Але якщо вас цікавить тестування, можливо розгляньте варіант переорієнтацію на Automation QA, оскількість більшість компапній стараються, де це можливо, перейти на автоматизацію процесів
- також мене зацікавило що серед 13 тисяч опитувальних - BI Analyst всього 24, а оскільки зростання обсягів даних створює потребу у висококваліфікованих аналітиках, і навіть не дуже великі компанії можуть мати великі обсяги данних і потребувати спеціалістів для їх обробки, то цей напрям має великий потенціал.
- Також прослідковується популярність аутсорсингу, що може свідчити про ефективність українських команд у роботі з міжнародними клієнтами.
- Найчастосний стаж роботи - 3 роки, це типовий вік для мідлів, що корелює з найпоширенішим тайтлом у даних.
- JavaScript, Python і Java домінують, що показує попит на web-розробку та універсальні мови.

Український IT-ринок демонструє високий рівень стійкості навіть у складних умовах. Ринок фокусується на Software Engineering, QA та аутсорсингу, але має перспективні ніші, такі як Al Programming та BI Analytics. Активність молодих спеціалістів із середнім стажем роботи (3 роки) свідчить про динамічний розвиток галузі.

10. Виведіть медіанну (медіана - це значення, менше якого є рівно 50% вибірки і більше якого є теж рівно 50% вибірки)
ЗП (Зарплата у $$$) для кожного рівня знання англійської. Які висновки можна зробити?

In [187]:
median_salary = df.groupby('Знання англійської мови')['Зарплата у $$$ за місяць, лише ставка після сплати податків'].median()
median_salary

Знання англійської мови
Advanced              3500.0
Elementary            1155.0
Intermediate          2112.5
Pre-Intermediate      1408.0
Upper-Intermediate    3300.0
Name: Зарплата у $$$ за місяць, лише ставка після сплати податків, dtype: float64

##### Висновок: 
Чим вище рівень знання англійської тим вище середня зп.

11. Є така штука як сортування значень категоріальної змінної. Нам це зараз стане в нагоді, адже рівні англійської можна посортувати.
Давайте відсотруємо рівні англ. і тайтли. Просто запустіть код нижче і проаналізуйте, що він робить.

In [191]:
df['Знання англійської мови'] = pd.Categorical(df['Знання англійської мови'],
                      categories=['Elementary','Pre-Intermediate','Intermediate','Upper-Intermediate','Advanced'],
                      ordered=True)

df['Ваш тайтл'] = pd.Categorical(df['Ваш тайтл'],
                      categories=['Intern/Trainee', 'Junior', 'Немає тайтлу', 'Middle', 'Senior', 'Tech Lead', 'Team Lead', 'Architect', 'Head', 'Manager'],
                      ordered=True)

Оскільки значення в колонках 'Знання англійської мови' та 'Ваш тайтл' мають певні визначенні значення, то з метою зменшення використовуваної пам'яті ми перетворюємо ці колонки в категоріальний тип данних. 

12. Зробіть зводну таблицю, де стовпці - це 'Ваш тайтл', рядки - 'Знання англійської мови', а значення - медіанна ЗП. Які висновки можна зробити?
Тут ще для зручності можна зробити градієнтне забарвлення і відформатувати значення для легшого перегляду.

In [259]:
# Створюємо зведену таблицю (pivot table)
table = pd.pivot_table(
    df, 
    index='Знання англійської мови',  
    columns='Ваш тайтл',             
    values='Зарплата у $$$ за місяць, лише ставка після сплати податків',  
    aggfunc='median'                 
)

# Форматуємо значення у зведеній таблиці та додаємо градієнтне забарвлення
styled_pivot_table = (
    table.style
    .format("{:.2f}")               
    .background_gradient(cmap="RdYlBu")  
)

# Відображаємо стилізовану зведену таблицю
styled_pivot_table

Співробітники з тайтлом “Архітектор” володіють англійською на рівні не нижче Pre-Intermediate, причому найвища середня зарплата спостерігається при рівні Advanced. Цікаво, що Tech Lead зі знанням англійської на рівні Elementary отримує вищу середню зарплату, ніж ті, хто має рівень Pre-Intermediate. Аналогічна тенденція спостерігається серед співробітників на позиції Intern/Trainee.

Така тенденція може свідчити про кілька факторів:

	1.	Досвід важливіший за рівень англійської: Співробітники з низьким рівнем мови можуть мати значний професійний досвід, що компенсує недостатні знання англійської.
	2.	Внутрішній ринок: Вони можуть працювати на локальних проектах або з клієнтами, які не потребують високого рівня англійської.
	3.	Невелика вибірка: Для рівня Elementary може бути мало даних, що викривляє середнє значення.
	4.	Складні проєкти: Навіть на рівні Intern/Trainee, низький рівень мови не виключає залучення до специфічних, високооплачуваних проєктів.


13. Виведіть результат методу describe() по колонцы розміру ЗП для топ 5 за частотою зустрічання в даних областей (В якій області ви зараз живете?). Зробіть висновки, де вигідніше жити з точки зору ЗП і чи суттєва різниця на Ваш погляд для прийняття рішення про переїзд?

In [693]:
top_regions_frequency = (
    df['В якій області ви зараз живете?']
    .value_counts()
    .head(5)
    .index
)

# Фільтруємо дані, залишаючи лише записи з цих регіонів
filtered_df = df[df['В якій області ви зараз живете?'].isin(top_regions_frequency)]

# Обчислюємо описову статистику зарплат для кожного регіону
result = filtered_df.groupby('В якій області ви зараз живете?')[
    'Зарплата у $$$ за місяць, лише ставка після сплати податків'
].describe()

# Сортуємо результат за кількістю записів (count) у спадному порядку
sorted_result = result.sort_values(by='count', ascending=False)

# Відображаємо відсортовану таблицю з результатами
display(sorted_result)

count    mean     std    min     25%  \
В якій області ви зараз живете?                                          
Київ чи область                 4408.00 3141.83 2039.90  80.00 1500.00   
Львів чи область                2351.00 3122.16 2103.36 150.00 1375.00   
Дніпро чи область                751.00 2582.58 1864.71 100.00 1025.00   
Івано-Франківськ чи область      478.00 2705.26 1905.67 200.00 1200.00   
Вінниця чи область               446.00 2878.97 1894.53 100.00 1300.00   

                                    50%     75%      max  
В якій області ви зараз живете?                           
Київ чи область                 2800.00 4500.00 16000.00  
Львів чи область                2700.00 4500.00 14000.00  
Дніпро чи область               2000.00 3600.00 12000.00  
Івано-Франківськ чи область     2372.50 3775.00 12436.00  
Вінниця чи область              2606.00 4087.50 11675.00

- Київ та область мають найвищий середній рівень зарплати — 3141.83, при цьому максимальна зарплата досягає 16,000. Це свідчить про те, що в Києві є найбільше можливостей для отримання високої зарплати, що відповідає статусу столиці з високим рівнем розвитку різних галузей.
- Львів та область займають другу позицію за середньою зарплатою — 3122.16, хоча максимальна зарплата (14,000) дещо нижча, ніж у Києві.

Прогноз щодо переїзду: Якщо  основне завдання — це максимізація доходу, то переїзд до Києва чи Львова може бути вигідним з огляду на високий середній рівень зарплати. Однак варто враховувати і інші фактори, як вартість життя, умови праці, доступність житла, наявність робочих місць у вашій галузі. Переїзд до регіонів з нижчою середньою зарплатою, таких як Дніпро чи Вінниця, може бути вигідним за умови оптимізації витрат і наявності гідних можливостей для кар’єрного росту.

14. Як зростає зарплата з роками? Для цього нам треба відформатувати колонку "Загальний стаж роботи за спеціальністю" аналогічно до того, як ми це робили для колонки 'Знання англійської мови' і потім вивести медіанні значення для кількості років стажу.

In [695]:
# Встановлення порядку для 'Загальний стаж роботи за спеціальністю'
order = [
    "Менше як 3 місяці", "3 місяці", "Пів року", "1 рік", "1,5 року", "2 роки",
    "3 роки", "4 роки", "5 років", "6 років", "7 років", "8 років", "9 років",
    "10 років", "11 років", "12 років", "13 років", "14 років", "15 і більше років"
]

# Перетворюємо на категоріальний тип даних колонку 'Загальний стаж роботи за спеціальністю'
df['Загальний стаж роботи за спеціальністю'] = pd.Categorical(
    df['Загальний стаж роботи за спеціальністю'], categories=order, ordered=True
)

# Обчислюємо медіанну зарплату залежно від загального стажу роботи за спеціальністю
salary_increase = df.groupby('Загальний стаж роботи за спеціальністю')[
    'Зарплата у $$$ за місяць, лише ставка після сплати податків'
].median()

# Відображаємо результати
display(salary_increase)

Загальний стаж роботи за спеціальністю
Менше як 3 місяці    800.00
3 місяці             525.00
Пів року             700.00
1 рік                950.00
1,5 року            1200.00
2 роки              1700.00
3 роки              2450.00
4 роки              3000.00
5 років             3500.00
6 років             4000.00
7 років             4250.00
8 років             4700.00
9 років             4500.00
10 років            4600.00
11 років            5000.00
12 років            4750.00
13 років            4500.00
14 років            5050.00
15 і більше років   4500.00
Name: Зарплата у $$$ за місяць, лише ставка після сплати податків, dtype: float64

Зарплата відповідно до стажу зростає пропорційно: чим вищій стаж - тим вища заробітна плата. Винятком в нашій вибірці є стаж 8 років де медіанна з/п становить 4700 що є вищім ніж при стажі 9-11 років. Можливо це пов'язано з тим, що співробітники зі стажем 8 років, ймовірно, займають посади з вищими обов’язками чи рівнем кваліфікації, отримуючи вищу оплату. Також це може свідчити про обмежену вибірку для стажів 9-12 років або нерівномірний розподіл посад у цих категоріях. Додатковий аналіз даних за посадами та галузями може допомогти уточнити причину цієї аномалії.

Далі тут можна зробити багато аналогічних агрегацій, проаналізувати, в який час люди заповнювали форму, хто більше отримує, Business Analyst, чи Data Analyst. Я вас заохочувати не стримувати свій інтерес і покапати ці дані далі :)

In [376]:
# Створюємо зведену таблицю для аналізу медіанної зарплати залежно від стажу та посади
table_years_title_salary = pd.pivot_table(
    df,
    index='Загальний стаж роботи за спеціальністю',  
    columns='Оберіть вашу посаду',  
    values='Зарплата у $$$ за місяць, лише ставка після сплати податків',  
    aggfunc='median'  
)

# Додаємо стиль: форматування значень до 2-х знаків після коми та градієнтне забарвлення
styled_table_years = table_years_title_salary.style.format('{:.2f}').background_gradient(cmap='coolwarm')

# Відображаємо стилізовану таблицю
styled_table_years

Згідно зведеної таблиці вище, цікаво прослідковується що в спеціалізації DBA стаж роботи від 2 років і вище, що може свідчити про доволі такі складні задачі в сфері адміністрування базами і новачків дуже мало в цій сфері. А в Customer Success навпаки більше 10-11 років ніхто не затримується і можливо після такого досвіду вже змінює спеціалізацію на Project/Product/Program Manager що вже перспективніше в розрізі розвитку зарплати та карьєрних можливостей.
Загалом розкид зростання зарплат до стажу роботи є не зовсім однорідний, що може сказати що все таки навички і вміння брати відповідальність і вирішувати складні задачі цінується більше ніж просто цифра у стажі роботи. 

Оскільки мене цікавить напрямок розвитку Data Analyst, Data Science, Machine Learning, Big Data, то подальший аналіз проведу для цих груп напрямку спеціалізації. 

In [510]:
# Створюємо зведену таблицю для аналізу гендерного розподілення за посадами
pivot_table_analytics = pd.pivot_table(
    df, 
    index= 'Ваша стать',
    columns = 'Ваша посада4', 
    aggfunc='size',
)
pivot_table_science = pd.pivot_table(
    df, 
    index= 'Ваша стать',
    columns = 'Ваша посада5',
    aggfunc='size',
)

# Об'єднання двох півот-таблиць
pivot_table_combined = pd.concat([pivot_table_analytics, pivot_table_science], axis=1)

# Виведення результату
display(pivot_table_combined)

BI Analyst  Business Analyst  Data Analyst  Product Analyst  \
Ваша стать                                                                
Жінка                8               140            37               11   
Чоловік             16               149            69               23   

            System Analyst  BI Engineer  Computer Vision Engineer  \
Ваша стать                                                          
Жінка                   14            4                         1   
Чоловік                 18            5                        20   

            Data Engineer / Big Data Engineer  Data Scientist  \
Ваша стать                                                      
Жінка                                      16              18   
Чоловік                                    88              75   

            Machine Learning Engineer  Research Engineer  
Ваша стать                                                
Жінка                               8                  1  
Чоловік                            38                 14

In [512]:
# Створюємо зведену таблицю для аналізу використання програмування на позиціях аналітиків та дата саєнтистів
pivot_analytics_programming = pd.pivot_table(
    df, 
    index= 'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
    columns = 'Ваша посада4', 
    aggfunc='size',
)
pivot_science_programming = pd.pivot_table(
    df, 
    index= 'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
    columns = 'Ваша посада5',
    aggfunc='size',
)

# Об'єднання двох півот-таблиць
pivot_programming_combined = pd.concat([pivot_analytics_programming, pivot_science_programming], axis=1)

# Виведення результату
display(pivot_programming_combined)

BI Analyst  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?               
Ні, не використовую                                                                  6   
Так, використовую                                                                   18   

                                                                            Business Analyst  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                     
Ні, не використовую                                                                      256   
Так, використовую                                                                         32   

                                                                            Data Analyst  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                 
Ні, не використовую                                                                   17   
Так, використовую                                                                     89   

                                                                            Product Analyst  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                    
Ні, не використовую                                                                       7   
Так, використовую                                                                        27   

                                                                            System Analyst  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                   
Ні, не використовую                                                                     20   
Так, використовую                                                                       12   

                                                                            BI Engineer  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                
Ні, не використовую                                                                1.00   
Так, використовую                                                                  8.00   

                                                                            Computer Vision Engineer  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                             
Ні, не використовую                                                                              NaN   
Так, використовую                                                                              21.00   

                                                                            Data Engineer / Big Data Engineer  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                                      
Ні, не використовую                                                                                      4.00   
Так, використовую                                                                                      100.00   

                                                                            Data Scientist  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                   
Ні, не використовую                                                                   1.00   
Так, використовую                                                                    92.00   

                                                                            Machine Learning Engineer  \
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?                              
Ні, не використовую                                                                               NaN   
Так, використовую                                                                               46.00   

                                                                            Research Engineer  
Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?

In [671]:
# Будуємо зведену таблицю середньої заробітної плати за стажем роботи для групи аналітиків з категорії 'Ваша посада4' 
salary_years = pd.pivot_table(
                            df, 
                            index = 'Загальний стаж роботи за спеціальністю', 
                            columns = 'Ваша посада4',
                            values = 'Зарплата у $$$ за місяць, лише ставка після сплати податків',
                            aggfunc = 'median'
                            )
salary_years

Ваша посада4                            BI Analyst  Business Analyst  \
Загальний стаж роботи за спеціальністю                                 
Менше як 3 місяці                           700.00            550.00   
3 місяці                                    600.00            700.00   
Пів року                                       NaN            850.00   
1 рік                                      1100.00           1300.00   
1,5 року                                   1000.00           1500.00   
2 роки                                     3500.00           2000.00   
3 роки                                     1950.00           2350.00   
4 роки                                     3205.00           2900.00   
5 років                                    1600.00           3150.00   
6 років                                    1350.00           4050.00   
7 років                                        NaN           3500.00   
8 років                                        NaN           4200.00   
9 років                                    4000.00           4550.00   
10 років                                   3800.00           4250.00   
11 років                                       NaN           4500.00   
12 років                                       NaN           1825.00   
13 років                                       NaN           4300.00   
14 років                                       NaN           3600.00   
15 і більше років                              NaN           2500.00   

Ваша посада4                            Data Analyst  Product Analyst  \
Загальний стаж роботи за спеціальністю                                  
Менше як 3 місяці                            1000.00           950.00   
3 місяці                                      460.00           550.00   
Пів року                                      600.00           850.00   
1 рік                                        1250.00           950.00   
1,5 року                                     1000.00          1100.00   
2 роки                                       1550.00          1255.00   
3 роки                                       2200.00          2775.00   
4 роки                                       2000.00          3250.00   
5 років                                      2575.00          3250.00   
6 років                                      2750.00              NaN   
7 років                                      2560.00          4250.00   
8 років                                      4800.00              NaN   
9 років                                      2650.00              NaN   
10 років                                     3000.00          2500.00   
11 років                                         NaN              NaN   
12 років                                         NaN              NaN   
13 років                                         NaN              NaN   
14 років                                         NaN              NaN   
15 і більше років                            2350.00              NaN   

Ваша посада4                            System Analyst  
Загальний стаж роботи за спеціальністю                  
Менше як 3 місяці                                  NaN  
3 місяці                                           NaN  
Пів року                                        950.00  
1 рік                                           775.00  
1,5 року                                        500.00  
2 роки                                             NaN  
3 роки                                         2200.00  
4 роки                                         3150.00  
5 років                                        3000.00  
6 років                                        1089.00  
7 років                                        2000.00  
8 років                                        4100.00  
9 років                                        3700.00  
10 років                                           NaN  
11 років                                 

In [709]:
# Групуємо за посадами і для кожної групи обираємо топ 4 найпоширеніших тайтлів для групи аналітиків 'Ваша посада4' 
top_titles_by_position = df.groupby('Ваша посада4')['Ваш тайтл'].value_counts().groupby(level=0).head(4)

display(top_titles_by_position)

Ваша посада4      Ваш тайтл     
BI Analyst        Junior              9
                  Middle              7
                  Senior              4
                  Немає тайтлу        3
Business Analyst  Middle            116
                  Senior             68
                  Junior             46
                  Немає тайтлу       30
Data Analyst      Junior             33
                  Middle             33
                  Немає тайтлу       20
                  Senior             12
Product Analyst   Middle             11
                  Senior              8
                  Junior              7
                  Intern/Trainee      2
System Analyst    Middle             10
                  Немає тайтлу        7
                  Junior              6
                  Senior              6
Name: Ваш тайтл, dtype: int64

In [703]:
# Будуємо зведену таблицю середньої заробітної плати за стажем роботи для групи аналітиків з категорії 'Ваша посада5' 
salary_years_science = pd.pivot_table(
                            df, 
                            index = 'Загальний стаж роботи за спеціальністю', 
                            columns = 'Ваша посада5',
                            values = 'Зарплата у $$$ за місяць, лише ставка після сплати податків',
                            aggfunc = 'median'
                            )
salary_years_science

Ваша посада5                            BI Engineer  Computer Vision Engineer  \
Загальний стаж роботи за спеціальністю                                          
Менше як 3 місяці                               NaN                       NaN   
3 місяці                                        NaN                       NaN   
Пів року                                        NaN                       NaN   
1 рік                                       1500.00                    740.00   
1,5 року                                    2000.00                       NaN   
2 роки                                       800.00                   1600.00   
3 роки                                          NaN                   3050.00   
4 роки                                      3800.00                   3800.00   
5 років                                         NaN                   4930.00   
6 років                                     1800.00                   5500.00   
7 років                                         NaN                       NaN   
8 років                                     3225.00                       NaN   
9 років                                         NaN                   6500.00   
10 років                                    5050.00                       NaN   
11 років                                        NaN                       NaN   
12 років                                        NaN                       NaN   
13 років                                    5400.00                       NaN   
14 років                                        NaN                       NaN   
15 і більше років                               NaN                       NaN   

Ваша посада5                            Data Engineer / Big Data Engineer  \
Загальний стаж роботи за спеціальністю                                      
Менше як 3 місяці                                                  267.50   
3 місяці                                                           250.00   
Пів року                                                          1300.00   
1 рік                                                              900.00   
1,5 року                                                          1715.00   
2 роки                                                            3000.00   
3 роки                                                            4200.00   
4 роки                                                            5466.00   
5 років                                                           5500.00   
6 років                                                           5500.00   
7 років                                                           6000.00   
8 років                                                           6880.00   
9 років                                                           4685.00   
10 років                                                          5000.00   
11 років                                                              NaN   
12 років                                                          6200.00   
13 років                                                          6800.00   
14 років                                                          4270.00   
15 і більше років                                                 5200.00   

Ваша посада5                            Data Scientist  \
Загальний стаж роботи за спеціальністю                   
Менше як 3 місяці                               800.00   
3 місяці                                           NaN   
Пів року                                        725.00   
1 рік                                          1100.00   
1,5 року                                       1500.00   
2 роки                                         2000.00   
3 роки                                         2945.00   
4 роки                                         3500.00   
5 років                                        4500.00   
6 років                                   

In [713]:
# Групуємо за посадами і для кожної групи обираємо топ 4 найпоширеніших тайтлів для групи саєнтистів 'Ваша посада5' 
top_titles_by_position_science = df.groupby('Ваша посада5')['Ваш тайтл'].value_counts().groupby(level=0).head(4)

display(top_titles_by_position_science)

Ваша посада5                       Ваш тайтл     
BI Engineer                        Middle             3
                                   Senior             3
                                   Junior             1
                                   Немає тайтлу       1
Computer Vision Engineer           Junior             6
                                   Middle             6
                                   Senior             3
                                   Tech Lead          3
Data Engineer / Big Data Engineer  Middle            34
                                   Senior            32
                                   Junior            19
                                   Немає тайтлу       7
Data Scientist                     Middle            29
                                   Junior            21
                                   Немає тайтлу      18
                                   Senior            15
Machine Learning Engineer          Junior            1

### Висновок:
При аналізі двох категорій напрямків Data Analyst та Data Science я бачу наступне:
    • Гендерна нерівність: У напрямках Data Analyst і Data Science простежується гендерна нерівність. У Data Analyst різниця між чоловіками та жінками менш значна, проте у Data Science кількість чоловіків у 4 рази перевищує кількість жінок. Це може бути зумовлено або упередженим ставленням під час найму, або страхом жінок перед можливими складнощами в цій галузі.
	•	Використання програмування: У більшості напрямків (особливо Data Science, BI Engineer, Machine Learning Engineer) активно використовуються мови програмування. Винятком є напрямок Business Analyst, де знання кодування рідко потрібне. Це пояснюється набором інструментів, які вони використовують у роботі (таблиці, презентації, BI-інструменти).
	•	Популярність позицій: Найбільша кількість спеціалістів у категорії Business Analyst мають тайтл Middle (116 осіб). Це свідчить про відносно легший поріг входу в професію, а також про зрозуміліші інструменти, порівняно зі складнішими позиціями, як-от Machine Learning Engineer чи BI Engineer.
	•	Рівень зарплат: У напрямках Data Science та ML розмах і максимальні рівні зарплат значно вищі, ніж у Data Analyst. Це підтверджує вищий рівень задач і складність роботи в цих напрямах, що зумовлює більший попит на таких спеціалістів.

##### Рекомендація для старту: Для початку кар’єри рекомендую обирати напрямок аналітики, оскільки він є зрозумілішим і доступнішим для освоєння. У подальшому можна переходити до спеціалізації в ML або Data Science, де відкриваються нові можливості та вищий рівень доходу.